sql: add right string function

aljoscha opened the pull request:
CHAR columns don't work proberly, see https://github.com/MaterializeInc/materialize/issues/589. I added commented tests that I would remove before merging.

Closes #5328

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/5966)
<!-- Reviewable:end -->

aljoscha commented on src/sql/src/func.rs:
```diff
@@ -1572,6 +1572,9 @@ lazy_static! {
                 params!(String, String, String) => VariadicFunc::Replace, 2087;
             },
+            "right" => Scalar {
+                params!(String, Int64) => BinaryFunc::Right, 42;
```
This number is made up because I didn't know what the number should be at that time.

JLDLaughlin replied:
42 was a good choice 😄

frankmcsherry commented on src/expr/src/scalar/func.rs:
```diff
@@ -4210,6 +4214,33 @@ fn trim_whitespace<'a>(a: Datum<'a>) -> Datum<'a> {
+        // if we went to far we just take the whole string
+        byte_indices.next().unwrap_or(0)
+    } else {
+        let n = -n;
```
Minor, but this could still be negative for `i64::MIN`. The `as usize` later on will do something, probably not as bad as UB, but did want to flag!

aljoscha replied:
I would use `i64::abs()`, which leaves it at `i64::MIN`. Or I could convert to `i128` before to make sure it always works, but it seems overkill. WDYT?

frankmcsherry replied:
Not too worried, just ideally avoiding crashing or anything like that.

JLDLaughlin commented on src/expr/src/scalar/func.rs:
```diff
@@ -4210,6 +4214,33 @@ fn trim_whitespace<'a>(a: Datum<'a>) -> Datum<'a> {
+    let string: &'a str = a.unwrap_str();
+    let n = b.unwrap_int64();
+
+    // the index of every char in bytes
```
The functions in this file are pretty sparsely documented (unless there's something particularly tricky), so I think you can remove the inline comments in `right`.

JLDLaughlin commented on src/expr/src/scalar/func.rs:
```diff
@@ -4210,6 +4214,33 @@ fn trim_whitespace<'a>(a: Datum<'a>) -> Datum<'a> {
+    let n = b.unwrap_int64();
+
+    // the index of every char in bytes
+    let byte_indices = string.char_indices().map(|(i, _)| i);
```
This chunk of code might be a bit clearer with direct index math instead of having to do the `rev().skip()` work. It would look something like:
```
    let start_index = if n >= 0 {
        if n >= string_len {
            0
        } else {
            string_len - n
        }
    } else {
        if n < 0 {
            string_len
        } else {
            n.abs()
        }
    };
```

aljoscha replied:
I find this one quite tricky. If I want to get `string_len` I need to do a `string.chars().count()`, which is `O(string_len)`.  And to then convert the `start_index` (in "char space") to a byte index I again have to traverse the iterator and sum up the bytes.

That's why I chose the somewhat complicated-looking reverse iteration. This way I only need to step `n` steps from the back, and I end up directly at the right byte index.

My solution isn't correct yet, as the test you suggested shows, but I could fix it. 😅

What do you think?

aljoscha replied:
I could be quite wrong on the string iteration/length business. Maybe I drank too much of the Rust-Iterators-are-cool Kool-Aid... 😅

aljoscha replied:
I pushed a change that makes it slightly less Iterator-y. I also added error handling with `usize::try_from()` but from looking at how the code is now, I don't think it can ever fail.

JLDLaughlin replied:
You were totally right! It turns out [`.len()`](https://doc.rust-lang.org/std/string/struct.String.html#method.len) returns the length of the `String` in bytes. Since chars/graphemes can be larger than a single byte, the index math I was suggesting would be wrong in these cases.

JLDLaughlin commented on test/sqllogictest/string.slt:
```diff
@@ -863,3 +863,94 @@ SELECT ''::pg_catalog.CHAR(0)
+(empty)
+
+query TT rowsort
+SELECT right(vccol, smicol), right(vccol, intcol) FROM righttest
```
Could you add a test like `right(vccol, 0)`? I believe your code will currently return the entire string, but this is what I get in Postgres:
```
jessicalaughlin=# select right('hiya', 0);
 right
-------

(1 row)
```

aljoscha replied:
Good catch! I'll add a test.

JLDLaughlin commented on src/expr/src/scalar/func.rs:
```diff
@@ -4210,6 +4214,33 @@ fn trim_whitespace<'a>(a: Datum<'a>) -> Datum<'a> {
+
+    let start_in_bytes = if n > 0 {
+        // pop off the last n-1 elements
+        let mut byte_indices = byte_indices.rev().skip((n - 1) as usize);
```
It's generally better to use Rust's [`TryFrom` Trait](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) instead of `as` to perform these conversions (and bubble up any errors). We don't do this everywhere ([but we want to!](https://materializeinc.slack.com/archives/CM7ATT65S/p1614821217253300)). This would look like:
```
usize::try_from(n-1)
```
There are other `try_from` calls in this file for reference!

frankmcsherry replied:
Another thing to do, if you have the spirit, is to track down what postgres does in this corner case. We'd error with `try_from`, but it could be that postgres does something specific like correctly handle strings of ridonkulous length.

aljoscha replied:
Thanks for being so thorough with me! 🍵 I'm still learning most of this Rust business.

JLDLaughlin commented on src/expr/src/scalar/func.rs:
```diff
@@ -4210,6 +4214,30 @@ fn trim_whitespace<'a>(a: Datum<'a>) -> Datum<'a> {
+    let string: &'a str = a.unwrap_str();
+    let n = b.unwrap_int64();
+
+    let mut byte_indices = string.char_indices().map(|(i, _)| i);
```
`byte_indices` is a bit misleading here, could you rename this to something like `char_indices` instead? (chars can span more than one byte!)

JLDLaughlin commented on test/sqllogictest/string.slt:
```diff
@@ -863,3 +863,105 @@ SELECT ''::pg_catalog.CHAR(0)
 query error length for type char must be within \[1-10485760\], have 10485761
 SELECT ''::pg_catalog.CHAR(10485761)
+
+### right ###
```
Sorry, I only figured this out after the first review/talking to @quodlibetor, but could you also add a test for [combining characters](https://en.wikipedia.org/wiki/Combining_character#:~:text=In%20digital%20typography%2C%20combining%20characters,marks%20(including%20combining%20accents).)? We only need to match Postgres' behavior here (which is a bit wonky):
```
jessicalaughlin=# select right('\u0930\u0903\u0951', 0);
 right
-------

(1 row)

jessicalaughlin=# select right('\u0930\u0903\u0951', 1);
 right
-------
 1
(1 row)

jessicalaughlin=# select right('\u0930\u0903\u0951', -1);
       right
-------------------
 u0930\u0903\u0951
```

aljoscha replied:
adding a test!

quodlibetor commented on test/sqllogictest/string.slt:
```diff
@@ -956,12 +956,35 @@ NULL    不释手
+----
+ः॑
+
+# this is exavtly the same as above, but using unicode escapes
```
minor typo: `exavtly` -> `exactly`

aljoscha replied:
had to be in this word... 🙈

philip-stoev reviewed (approved):
Thank you for the tests. I believe you have added all the corner cases I could think of, nulls, empty strings, negative offsets . So nothing else for me to add here.

JLDLaughlin commented on test/sqllogictest/string.slt:
```diff
@@ -863,3 +863,128 @@ SELECT ''::pg_catalog.CHAR(0)
+----
+NULL
+
+# combining characters
```
This is awesome, thank you!

frankmcsherry reviewed (approved):
Looks good to me. Only comment is that we might error on `i64::MIN` whereas pg might not. Idk, and not too worried, but wanted to mention it!

aljoscha commented:
I checked these:
```
aljoscha=# select right('hello', -9223372036854775808); // i64::MIN
ERROR:  function right(unknown, bigint) does not exist
LINE 1: select right('hello', -9223372036854775808);
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
aljoscha=# select right('hello', -9223372036854775807); // i64::MIN + 1
ERROR:  function right(unknown, bigint) does not exist
LINE 1: select right('hello', -9223372036854775807);
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

So `i64` is not supported at all.

```
aljoscha=# select right('hello', -2147483648);  // i32::MIN
 right
-------
 hello
(1 row)

aljoscha=# select right('hello', -2147483647); // i32::MIN + 1
 right
-------

(1 row)
```

Wraps around on `i32::MIN`, which we don't do because we use `i64`. So there is a difference in behaviour on `i32::MIN`.

aljoscha commented:
Well, that's a bit surprising:
```
    SELECT right('hello', -2147483648)
    PlanFailure:test/sqllogictest/string.slt:975:
    db error: ERROR: Cannot call function right(unknown, numeric(38,0)): arguments cannot be implicitly cast to any implementation's parameters; try providing explicit casts: ERROR: Cannot call function right(unknown, numeric(38,0)): arguments cannot be implicitly cast to any implementation's parameters; try providing explicit casts
    ----
    SELECT right('hello', -9223372036854775808)
    PlanFailure:test/sqllogictest/string.slt:987:
    db error: ERROR: Cannot call function right(unknown, numeric(38,0)): arguments cannot be implicitly cast to any implementation's parameters; try providing explicit casts: ERROR: Cannot call function right(unknown, numeric(38,0)): arguments cannot be implicitly cast to any implementation's parameters; try providing explicit casts
    ----
    SELECT right('hello', -9223372036854775807)
    PlanFailure:test/sqllogictest/string.slt:993:
    db error: ERROR: Cannot call function right(unknown, numeric(38,0)): arguments cannot be implicitly cast to any implementation's parameters; try providing explicit casts: ERROR: Cannot call function right(unknown, numeric(38,0)): arguments cannot be implicitly cast to any implementation's parameters; try providing explicit casts
    ----
```

We fail on everything `< i32::MIN + 1`, even though we define `right` to have these types:
```
"right" => Scalar {
    params!(String, Int64) => BinaryFunc::Right, 3061;
},
```

aljoscha commented:
Fyi, I'm waiting on #6019 so that I can test the behavior for integer edge cases.

aljoscha commented:
I changed the behaviour slight, Postgres doesn't support `bigint` and has weird behaviour for `i32::MIN`. I added tests for that, see the second commit on top.

JLDLaughlin commented on src/expr/src/scalar/func.rs:
```diff
@@ -4282,6 +4286,34 @@ fn position<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
+        // nth from the back
+        byte_indices.rev().nth(n).unwrap_or(0)
+    } else if n == i32::MIN {
+        // this seems strange but Postgres behaves like this
```
👀 wild.
